pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/SparkSourceOffset.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.google.cloud.pubsublite.Partition;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public final class SparkSourceOffset
+    extends org.apache.spark.sql.sources.v2.reader.streaming.Offset {
+  private static final ObjectMapper objectMapper =
+      new ObjectMapper().configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
+
+  // Using a map to ensure unique partitions.
+  private final ImmutableMap<Partition, SparkPartitionOffset> partitionOffsetMap;
+
+  public SparkSourceOffset(Map<Partition, SparkPartitionOffset> map) {
+    validateMap(map);
+    this.partitionOffsetMap = ImmutableMap.copyOf(map);
+  }
+
+  private static void validateMap(Map<Partition, SparkPartitionOffset> map) {
+    map.forEach(
+        (k, v) -> {
+          assert Objects.equals(k, v.partition())
+              : "Key(Partition) and value(SparkPartitionOffset)'s partition don't match.";
+        });
+  }
+
+  public static SparkSourceOffset merge(SparkSourceOffset o1, SparkSourceOffset o2) {
+    Map<Partition, SparkPartitionOffset> result = new HashMap<>(o1.partitionOffsetMap);
+    o2.partitionOffsetMap.forEach(
+        (k, v) ->
+            result.merge(
+                k,
+                v,
+                (v1, v2) ->
+                    SparkPartitionOffset.builder()
+                        .partition(Partition.of(k.value()))
+                        .offset(Collections.max(ImmutableList.of(v1.offset(), v2.offset())))
+                        .build()));
+    return new SparkSourceOffset(result);
+  }
+
+  public static SparkSourceOffset merge(SparkPartitionOffset[] offsets) {
+    Map<Partition, SparkPartitionOffset> map = new HashMap<>();
+    for (SparkPartitionOffset po : offsets) {
+      assert !map.containsKey(po.partition()) : "Multiple PslPartitionOffset has same partition.";
+      map.put(
+          po.partition(),
+          SparkPartitionOffset.builder().partition(po.partition()).offset(po.offset()).build());
+    }
+    return new SparkSourceOffset(map);
+  }
+
+  @SuppressWarnings("unchecked")
+  public static SparkSourceOffset fromJson(String json) {
+    Map<String, Number> map;
+    try {
+      // TODO: Use TypeReference instead of Map.class, currently TypeReference breaks spark with
+      // java.lang.LinkageError: loader constraint violation: loader previously initiated loading
+      // for a different type.
+      map = objectMapper.readValue(json, Map.class);
```

Comment:
don't know how to align files but this is the only real change other than renaming.